numpy/ma/core.py

```diff
@@ -5079,24 +5007,66 @@ def mean(self, axis=None, dtype=None, out=None):
         1.5
 
         """
+        kwargs = {}
+        if keepdims is not np._NoValue:
+            kwargs['keepdims'] = keepdims
+
         if self._mask is nomask:
-            result = super(MaskedArray, self).mean(axis=axis, dtype=dtype)
-        else:
-            dsum = self.sum(axis=axis, dtype=dtype)
-            cnt = self.count(axis=axis)
-            if cnt.shape == () and (cnt == 0):
-                result = masked
+            ret = self._data.mean(axis, dtype, out, **kwargs)
+
+            to_mask = ~np.isfinite(ret)
+            if self.size == 0:
+                to_mask = True
+
+            if out is not None:
+                if isinstance(out, MaskedArray):
+                    out.__setmask__(to_mask if np.any(to_mask) else False)
+                return out
+
+            if isinstance(ret, ndarray):
+                ret = ret.view(type(self))
+                ret.__setmask__(to_mask if np.any(to_mask) else False)
+                return ret
+            else:
+                # otherwise it's a scalar
+                return ret if self.size else masked
+
+        # code largely copied from _mean in numpy/core/_methods.py
+
+        rcount = self.count(axis=axis, **kwargs)
+
+        # Cast bool, unsigned int, and int to float64 by default
+        if dtype is None and issubclass(self.dtype.type,
+                                        (np.integer, np.bool_)):
+            dtype = np.dtype('f8')
+
+        ret = self.sum(axis=axis, dtype=dtype, **kwargs)
+        if ret is masked:
+            return _return_float_masked(out)
+        with np.errstate(divide='ignore', invalid='ignore'):
+            if isinstance(ret, np.ndarray):
+                ret = np.true_divide(
+                        ret, rcount, out=ret, casting='unsafe', subok=False)
```

Comment:
So after trying it out, it seems that the reason for using `true_divide` instead of `/=` was to be able to do `casting='unsafe'`. The inplace operation does not allow unsafe casting:
```
>>> m = np.ones(4, dtype='l')
>>> m /= 2.0
TypeError: ufunc 'divide' output (typecode 'd') could not be coerced to provided
output parameter (typecode 'l') according to the casting rule ''same_kind''
```

I think the way out which preserves back-compatibility is to do
```
ret = ret / rcount
```
and then update treatment of the `out` parameter appropriately, even though it unfortunately creates an unecessary buffer and defeats the inplace benefit of the `out` parameter.